File: docs/part2/chapter20.py
# Chapter 20 뒷정리하기

# 테스트 전략 변경 : flag 확인에서 log 확인으로!
class TestCase:
  def __init__(self, name):
    self.name = name
  def setUp(self):
    pass
  def run(self):
    self.setUp()
    method = getattr(self, self.name)
    method()
    self.tearDown()
  def tearDown(self):
    pass

class WasRun(TestCase):
  def __init__(self, name):
    self.wasRun = None
    TestCase.__init__(self, name)
  def testMethod(self):
    self.wasRun = 1
    self.log= self.log + "testMethod "
  def setUp(self):
    self.wasRun = None
    self.wasSetUp = 1
    self.log = "setUp "
  def tearDown(self):
    self.log = self.log + "tearDown "

class TestCaseTest (TestCase) :
  def setUp(self) :
    self.test = WasRun("testMethod")
  def testTemplateMethod(self):
    test= WasRun("testMethod")
    test.run()
    assert("setUp testMethod tearDown " == test.log)

TestCaseTest("testTemplateMethod").run()

File: docs/part2/chapter20.js
class TestCase {
  constructor(name) {
    this.name = name;
  }
  setUp = () => {
    return;
  };
  run = () => {
    this.setUp();
    if (!this[this.name]) return;
    const method = this[this.name];
    method();
    this.tearDown();
  };

  tearDown = () => {
    return;
  };
}

class WasRun extends TestCase {
  constructor(name) {
    super(name);
    this.wasRun = 0;
  }

  testMethod = () => {
    this.wasRun = 1;
    this.log += "testMethod ";
  };

  setUp = () => {
    this.wasRun = 0;
    this.wasSetUp = 1;
    this.log = "setUp ";
  };

  tearDown = () => {
    this.log += "tearDown";
  };
}

class TestCaseTest extends TestCase {
  constructor(name) {
    super(name);
  }

  setUp = () => {
    this.test = new WasRun("testMethod");
  };
  testTemplateMethod = () => {
    this.setUp();
    this.test.run();
    console.log("setUp testMethod tearDown" === this.test.log);
  };
  // testRunning = () => {
  //   const test = new WasRun("testMethod");
  //   test.run();
  //   return test.wasRun;
  // };
  // testSetUp = () => {
  //   this.setUp();
  //   this.test.run();
  //   return this.test.wasSetUp;
  // };
}

const test = new TestCaseTest("testTemplateMethod");
test.run();
